crates/federate/src/worker.rs

```diff
@@ -166,203 +235,534 @@ impl InstanceWorker {
       self.save_and_send_state().await?;
       latest_id
     };
-    if id >= latest_id {
-      if id > latest_id {
-        tracing::error!(
-          "{}: last successful id {} is higher than latest id {} in database (did the db get cleared?)",
-          self.instance.domain,
-          id.0,
-          latest_id.0
-        );
+    Ok((last, latest_id))
+  }
+
+  async fn handle_send_results(&mut self) -> Result<(), anyhow::Error> {
+    let mut force_write = false;
+    let mut events = Vec::new();
+    // Wait for at least one event but if there's multiple handle them all.
+    // We need to listen to the cancel event here as well in order to prevent a hang on shutdown:
+    // If the SendRetryTask gets cancelled, it immediately exits without reporting any state.
+    // So if the worker is waiting for a send result and all SendRetryTask gets cancelled, this recv
+    // could hang indefinitely otherwise. The tasks will also drop their handle of
+    // report_send_result which would cause the recv_many method to return 0 elements, but since
+    // InstanceWorker holds a copy of the send result channel as well, that won't happen.
+    tokio::select! {
+      _ = self.receive_send_result.recv_many(&mut events, 1000) => {},
+      () = self.stop.cancelled() => {
+        tracing::debug!("cancelled worker loop while waiting for send results");
+        return Ok(());
       }
-      // no more work to be done, wait before rechecking
-      tokio::select! {
-        () = sleep(*WORK_FINISHED_RECHECK_DELAY) => {},
-        () = self.stop.cancelled() => {}
+    }
+    for event in events {
+      match event {
+        SendActivityResult::Success(s) => {
+          self.state.fail_count = 0;
+          self.in_flight -= 1;
+          if !s.was_skipped {
+            self.mark_instance_alive().await?;
+          }
+          self.successfuls.push(s);
+        }
+        SendActivityResult::Failure { fail_count, .. } => {
+          if fail_count > self.state.fail_count {
+            // override fail count - if multiple activities are currently sending this value may get
+            // conflicting info but that's fine
+            self.state.fail_count = fail_count;
```

Comment:
Just increment by one instead of using assignment. Then theres no need for `SendActivityResult::Failure.fail_count`

Reply:
This needs to be this way, all alternatives would be worse. The reason is that if 10 simultaneous requests fail within a 1s period, we don't want the next retry to be exponentially 2**10 s later.